File: checks/ruby/lib/section508/checks/skip_link.rb
# frozen_string_literal: true

module Section508
  module Checks
    # The first focusable control in the specimen jumps to unique content.
    class SkipLink
      def self.call(root)
        return [] unless root&.respond_to?(:css)

        focusable = focusable_controls(root)
        return ["There is no skip link."] if focusable.empty?

        first = focusable.first
        href = first["href"].to_s
        name = "#{first.text} #{first["aria-label"]}".gsub(/\s+/, " ").strip

        unless href.start_with?("#") && href.length >= 2 && /skip/i.match?(name)
          return ["The first control is not a skip link."]
        end

        id = href[1..]
        target = (Dom.element?(root) && root["id"] == id) ? root : Dom.by_id(root, id)
        return ["The skip link does not point at an element in this section."] unless target

        []
      end

      def self.focusable_controls(root)
        Dom.elements(root, "a[href], button, input, select, textarea, [tabindex]").select do |element|
          primary = native_focusable?(element)
          next false if !primary && element["tabindex"] == "-1"
          next false if Dom.has_attr?(element, "disabled")
          next false if Dom.has_attr?(element, "hidden")
          next false if element["aria-hidden"] == "true"

          style = element["style"].to_s.gsub(/\s+/, "")
          next false if /display:none/i.match?(style) || /visibility:hidden/i.match?(style)

          true
        end
      end

      def self.native_focusable?(element)
        name = element.name.upcase
        return true if %w[BUTTON INPUT SELECT TEXTAREA].include?(name)

        name == "A" && Dom.has_attr?(element, "href")
      end
    end
  end
end

File: checks/python/src/section508/checks/skip_link.py
import re

from section508.dom import SPACE, attr, attr_or, find_id, has_attr, select

SKIP = re.compile(r"skip", re.I)
HIDDEN_STYLE = re.compile(r"display:none", re.I)
VIS_HIDDEN = re.compile(r"visibility:hidden", re.I)


def check(root):
    if not root:
        return []

    focusable = focusable_controls(root)
    if not focusable:
        return ["There is no skip link."]

    first = focusable[0]
    href = attr(first, "href") or ""
    name = SPACE.sub(" ", f"{first.get_text() or ''} {attr_or(first, 'aria-label')}").strip()

    if not href.startswith("#") or len(href) < 2 or not SKIP.search(name):
        return ["The first control is not a skip link."]

    target_id = href[1:]
    target = find_id(root, target_id)
    if target is None:
        return ["The skip link does not point at an element in this section."]
    return []


def focusable_controls(root):
    found = []
    for element in select(
        root,
        'a[href], button, input, select, textarea, [tabindex]:not([tabindex="-1"])',
    ):
        if has_attr(element, "disabled"):
            continue
        if has_attr(element, "hidden"):
            continue
        if attr(element, "aria-hidden") == "true":
            continue
        style = (attr(element, "style") or "").replace(" ", "")
        if HIDDEN_STYLE.search(style) or VIS_HIDDEN.search(style):
            continue
        found.append(element)
    return found
